src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi

```diff
@@ -104,10 +109,67 @@ async def callback_start_batch(GrpcCallWrapper grpc_call_wrapper,
         wrapper.c_functor(), NULL)
 
     if error != GRPC_CALL_OK:
-        raise RuntimeError("Failed grpc_call_start_batch: {}".format(error))
+        raise CallbackStartBatchError("Failed grpc_call_start_batch: {}".format(error))
 
     await future
     cpython.Py_DECREF(wrapper)
     cdef grpc_event c_event
     # Tag.event must be called, otherwise messages won't be parsed from C
     batch_operation_tag.event(c_event)
+
+
+async def _receive_message(GrpcCallWrapper grpc_call_wrapper,
+                           object loop):
+    """Retrives parsed messages from C-Core.
+    
+    The messages maybe already in C-Core's buffer, so there isn't a 1-to-1
+    mapping between this and the underlying "socket.read()". Also, eventually,
+    this function will end with an EOF, which reads empty message.
+    """
+    cdef ReceiveMessageOperation receive_op = ReceiveMessageOperation(_EMPTY_FLAG)
+    cdef tuple ops = (receive_op,)
+    try:
+        await callback_start_batch(grpc_call_wrapper, ops, loop)
+    except CallbackStartBatchError as e:
+        # NOTE(lidiz) The receive message operation has two ways to indicate
+        # finish state : 1) returns empty message due to EOF; 2) fails inside
+        # the callback (e.g. cancelled).
+        #
+        # Since they all indicates finish, they are better be merged.
+        _LOGGER.exception(e)
+    return receive_op.message()
+
+
+async def _send_message(GrpcCallWrapper grpc_call_wrapper,
+                        bytes message,
+                        bint metadata_sent,
+                        object loop):
+    cdef SendMessageOperation op = SendMessageOperation(message, _EMPTY_FLAG)
+    cdef tuple ops
+    if metadata_sent:
+        ops = (op,)
+    else:
+        ops = (
+            # Initial metadata must be sent before first outbound message.
+            SendInitialMetadataOperation(None, _EMPTY_FLAG),
```

Comment:
Interesting quirk of the Core API.

Optional: We could also solve this problem by making Core more lenient and doing the equivalent of this in the case that the first `SendMessageOperation` comes without a preceding `SendInitialMetadataOperation`. Seems like this would be a most backwards compatible change.

Reply:
That is a doable solution with different tradeoffs. E.g. Core might need another layer of abstraction to insert certain operations, or the error handling path needs to indicate that `InitialMetadata` is sent automatically.